fake-koji/src/main/java/org/fakekoji/jobmanager/project/JDKProjectParser.java

```diff
@@ -381,4 +390,303 @@ void setProjectVariables(List<OToolVariable> variables) {
+        final Optional<VariantsConfig> configOptional = findVariants(configs, map);
+        if (configOptional.isPresent()) {
+            return configOptional.get();
+        } else {
+            final VariantsConfig config = new VariantsConfig(map, new HashSet<>());
+            configs.add(config);
+            return config;
+        }
+    }
+
+    private static Map<String, String> variantsMapToStringMap(Map<TaskVariant, TaskVariantValue> variants) {
+        return variants.entrySet()
+                .stream()
+                .collect(Collectors.toMap(e -> e.getKey().getId(), e -> e.getValue().getId()));
+    }
+
+    private static String validateJobField(final String label, final Object o1, final Object o2) {
+        return o1.equals(o2)
+                ? ""
+                : label + " " + o1.toString() + " doesn't match " + o2.toString() + '\n';
+    }
+
+    private static Function<TaskJob, Function<TaskJob, String>> getTaskJobValidator() {
+        return randomJob -> job -> validateJobField("project", job.getProjectName(), randomJob.getProjectName()) +
+                validateJobField("product", job.getProduct(), randomJob.getProduct()) +
+                validateJobField("jdk version", job.getJdkVersion(), randomJob.getJdkVersion()) +
+                validateJobField("build providers", job.getBuildProviders(), randomJob.getBuildProviders()) +
+                validateJobField("variables", job.getProjectVariables(), randomJob.getProjectVariables()) +
+                validateJobField("scripts root", job.getScriptsRoot(), randomJob.getScriptsRoot());
+    }
+
+    private static Function<TestJob, Function<TestJob, String>> getTestJobValidator() {
+        return randomJob -> {
+            final Function<TaskJob, String> taskJobValidator = getTaskJobValidator().apply(randomJob);
+            return job -> taskJobValidator.apply(job) +
+                    validateJobField(
+                            "project subpackage blacklist",
+                            job.getProjectSubpackageBlacklist(),
+                            randomJob.getProjectSubpackageBlacklist()
+                    ) +
+                    validateJobField(
+                            "project subpackage whitelist",
+                            job.getProjectSubpackageWhitelist(),
+                            randomJob.getProjectSubpackageWhitelist()
+                    );
+        };
+    }
+
+    private static <T extends Job> Result<T, String> getRandomJob(Set<T> jobs) {
+        final Optional<T> randomJobOptional = jobs.stream().findFirst();
+        return randomJobOptional
+                .<Result<T, String>>map(Result::ok)
+                .orElseGet(() -> Result.err("Set of jobs is empty"));
+    }
+
+    private static <T extends TaskJob> Result<T, String> validateJobs(Set<T> jobs, Function<T, Function<T, String>> validator) {
+        return getRandomJob(jobs).flatMap(randomJob -> {
```

Comment:
getRadnom job!??!!?!?

Reply:
yeah, it's a job that determines project name, url or subpackagelists

Reply:
As this is part of (in future) extraced code to new class, it very likely deserves some javadoc. it is very hard to read logic